Envest/single cell prediction nb

envest opened the pull request:
This PR explores results from using a kTSP prediction model to predict the subgroup of individual cells. Overall, it works!

For some samples, there is a mixture of subgroups predicted within the population of cells. The confidence level given to those predictions tends to be lower. This is visualized, along with a UMAP mapping of clusters and cell subgroup.

Your much greater experience using markdown notebooks could be a great benefit here, along with ideas for how to show these ideas more convincingly. Thanks!

envest commented:
👋 @jaclyn-taroni ready for review! I reformatted the `patchwork` plot a bit to get everything to show up nicely, and also added a plot that shows how prediction scores for G3 and G4 exist on a continuum between G3 and G4. Some cells are all G3, some are all G4, but most have signal from both. That feels like a supplemental figure to me, or maybe even simply left in the notebook.

Thanks for your review!

jaclyn-taroni commented on analysis_notebooks/test_pseudobulk_and_single_cells.Rmd:
````diff
@@ -0,0 +1,510 @@
+
+```{r}
+  
+return_umap_cluster <- function(sample_title) {
````
```suggestion
get_umap_coord_and_cluster <- function(sample_title) {
```

Here's why: if you told me you were deriving the cluster labels using UMAP instead of PCA, it would give me pause.

jaclyn-taroni commented on analysis_notebooks/test_pseudobulk_and_single_cells.Rmd:
```diff
@@ -0,0 +1,510 @@
+pseudobulk_plot_data_filepath <- file.path(plots_data_dir, "pseudobulk_test_predictions.tsv")
+pseudobulk_subgroups_plot_filepath <- file.path(plots_dir, "single_cell_pseudobulk_subgroups.pdf")
+percentage_subgroups_plot_filepath <- file.path(plots_dir, "single_cell_percentage_subgroups.pdf")
+confidence_umap_plot_filepath <- file.path(plots_dir, "single_cell_confidence_umap.pdf")
```
I have a violin plot for this one?

envest replied:
Yes -- resolved by removed "_umap"

jaclyn-taroni commented on analysis_notebooks/test_pseudobulk_and_single_cells.Rmd:
````diff
@@ -0,0 +1,510 @@
+
+```
+
+### Plot UMAPs with subgroup prediction
````
This header is the same as the next one, and I have a violin plot instead of a UMAP.

envest replied:
Updated to reflect it is UMAP with assigned clusters

envest replied:
File names resolved

jaclyn-taroni commented on analysis_notebooks/test_pseudobulk_and_single_cells.Rmd:
````diff
@@ -0,0 +1,510 @@
+confidence_umap_plot_object
+```
+
+### Plot UMAPs with subgroup prediction
````
This is cluster, not subgroup prediction. I still maintain that cluster label is not really that informative for these analyses.

envest replied:
Updated to reflect it is UMAP with assigned clusters -- agree on limited utility of assigned cluster labels

jaclyn-taroni commented on analysis_notebooks/test_pseudobulk_and_single_cells.Rmd:
```diff
@@ -0,0 +1,510 @@
+  ggplot2::theme(legend.position = "bottom",
+                 legend.direction = "horizontal")
+
+ggplot2::ggsave(filename = confidence_umap_plot_filepath,
```
Ah, here's the problem. I'm not sure what the filename variable should be here based on the output files at the beginning of the notebook.

envest replied:
File names resolved

jaclyn-taroni reviewed (commented):
Thanks for sending this over, @envest! My biggest concern here is how long the main notebook you’re adding here is – because it’s difficult for me (and possibly you!) to follow and verify correctness at its current length. (See the comments about the violin plot.) It’s also pretty light on comments.

Can the prediction and data preparation steps be split from the plotting steps? I don’t think you need a notebook to do the first part.  If you use a notebook for plotting, you’ll have more of an opportunity to summarize the results in the notebook. You do have a bit of that already, but it’s going to be very kind to future you for a project like this that’s going to have results trickling in over a long time span if those summaries are very solid.

I also think there’s an opportunity to potentially simplify how you’re detecting/setting model types for classifiers and using `purrr::map()` throughout. That's less "mission critical" in my opinion.

envest commented on 02-gather_data.R:
```diff
@@ -187,61 +187,68 @@ pseudobulk_metadata <- readr::read_tsv(pseudobulk_metadata_input_filepath,
+    pseudobulk_df <- pseudobulk_df |>
+      tibble::add_column("{sample_title}" := average_tpm_vector)
+
+  }
```
Yes that's right -- added a comment to clarify what's happening

envest commented on utils/single-cell.R:
```diff
@@ -104,10 +104,114 @@ perform_graph_clustering <- function(sce,
+
+  if ( model_type %in% c("ktsp", "rf") ) {
+
+    if ( model_type == "ktsp" & class(classifier) != "OnevsrestScheme_TSP" ) {
+
+      stop("Type of model specified by classifier parameter in test_single_cells() does not match model_type, which must be 'ktsp' or 'rf'.")
+
+    }
+
+    if ( model_type == "rf" & class(classifier) != "rule_based_RandomForest" ) {
+
+      stop("Type of model specified by classifier parameter in test_single_cells() does not match model_type, which must be 'ktsp' or 'rf'.")
+
+    }
```
Resolved by defining `model_type` based on the class of classifier, as suggested in another comment

envest commented on utils/single-cell.R:
```diff
@@ -104,10 +104,114 @@ perform_graph_clustering <- function(sce,
+  # read in single cell experiment object
+  sce_object <- readr::read_rds(sce_filepath)
+
+  # read in gene expression matrix
+  genex_df_this_sample <- sce_object@assays@data@listData[[1]]
```
Yes... it's just `SingleCellExperiment::counts()` 🤦

envest commented on utils/single-cell.R:
```diff
@@ -104,10 +104,114 @@ perform_graph_clustering <- function(sce,
+
+  if (length(sample_subgroup) != 1) {
+
+    stop("Wait -- sample accession is not unique in scRNA metadata")
```
_I think_ the purpose of the test here is to make sure there the sample accession is unique in the metadata... and using the length of `sample_subgroup` in the test feeds two birds with one scone. _Maybe._

Verbal processing for myself: filtering for a particular sample accession should result in a unique sample, and pulling that sample's subgroup should give a vector of length 1. If that vector is longer than 1, more than one row was returned by the filter. And the sample accession is not unique in the metadata.

I've added a comment to that effect and also made the error message more direct and less like what I wrote to amuse myself when including what was probably a debugging test in the first place 😄

envest commented on analysis_notebooks/test_pseudobulk_and_single_cells.Rmd:
```diff
@@ -0,0 +1,510 @@
+# Load libraries 
+source(here::here("utils/color_schemes.R"))
+source(here::here("utils/modeling.R"))
+source(here::here("utils/single-cell.R"))
```
Search/replaced "file.path" for "here::here"

envest commented on analysis_notebooks/test_pseudobulk_and_single_cells.Rmd:
````diff
@@ -0,0 +1,510 @@
+```{r}
+
+pseudobulk_subgroups_plot_object <- pseudobulk_plot_df |>
+  dplyr::filter(model_type == "kTSP (w)") |>
````
Resolved by faceting on kTSP and RF instead of only showing kTSP

envest commented:
@jaclyn-taroni Update: I've updated scripts to incorporate all suggestions, including:

- Split prediction tasks from plotting tasks
- Included RF model results
- Used model types as defined in the models, not hardcoded
- Fixed plot file names and section titles

What I have not done yet is go back to re-interpret results and reimagine how to illustrate our main points in the notebook. That is next!

envest commented:
@jaclyn-taroni You've already requested yourself. We discussed this review should cover technical changes, and we'll save interpretative / creative visualization changes for a later date, after we fix some aspects of modeling. Thanks!

jaclyn-taroni commented on predict/predict_pseudobulk_and_single_cells.R:
```diff
@@ -0,0 +1,164 @@
+                     dplyr::select(sample_accession,
+                                   subgroup),
+                   by = "sample_accession") |>
+  dplyr::mutate(model_type = dplyr::case_match(model_type,
+                                               "ktsp" ~ "kTSP (w)",
+                                               "rf" ~ "RF (w)"))
```
Checking my understanding: This should be fine since we're only using the weighted versions.

envest replied:
Yes fine for now re: #73 and #63 + #75. We can return to this point when more models are incorporated.

jaclyn-taroni commented on predict/predict_pseudobulk_and_single_cells.R:
```diff
@@ -0,0 +1,164 @@
+
+# Session info
+
+sessionInfo()
```
I don't think you need this unless you plan to write the output somewhere.

jaclyn-taroni commented on predict/predict_pseudobulk_and_single_cells.R:
```diff
@@ -0,0 +1,164 @@
+  dplyr::rowwise() |>
+  dplyr::mutate(max = max(G3, G4, SHH, WNT),
+                total = sum(G3, G4, SHH, WNT),
+                confidence = max/total,
```
This is related to a discussion we had at our meeting: I don't know if we want to do this for the RF models since we expect them to return probabilities. One could imagine a different visualization for the RF models (dodged barplot perhaps?).

envest replied:
Agree -- confidence won't be a useful metric for RF, if only because we don't use it! This code keeps G3, G4, SHH, and WNT columns, so we will still be able to show those when plotting RF in a style different from kTSP. I'm curious to see what metrics are most useful with medulloPackage.

jaclyn-taroni reviewed (commented):
I think this is close. I am of the opinion that we should return to the RF visualizations at a later time. Since we expect that model returns probabilities, I think we should show people probabilities instead of treating the output score the same way we treat the kTSP models. I would be satisfied for now if we skipped creating confidence scores for RF models and removing all RF plotting from the notebook.
